### DISCUSSION TITLE
Nullable reference type in explicit interface implementation, warning CS8618

### DISCUSSION THREAD

**Discussion by ilbrando** [dotnet/roslyn#59879] [Q&A] [answered]
Is it a bug that the compiler can't figure out that Bar does get initialized in the constructor?

In the explicit interface implementation I get a CS8618 warning: *Non-nullable IFoo.Bar must contain a non-null value when exiting constructor*
When using implicit implementation there are no warnings.

How is nullable reference types supposed to be used with explicit interface implementations?

```csharp
public interface IFoo
{
  string Bar { get; set; }
}

public class FooExplicit : IFoo
{
  public FooExplicit(string bar)
  {
    if (bar is null) throw new ArgumentNullException(nameof(bar));
    ((IFoo)this).Bar = bar;
  }
  string IFoo.Bar { get; set; }
}

public class FooImplicit : IFoo
{
  public FooImplicit(string bar)
  {
    if (bar is null) throw new ArgumentNullException(nameof(bar));
    Bar = bar;
  }
  public string Bar { get; set; }
}
```

Thanks for any help

**Comment by ilbrando** [comment, marked as answer]
I found a work around. Using a private field.
```csharp
public interface IFoo
{
  string Bar { get; set; }
}

public class FooExplicit : IFoo
{
  private readonly string _bar;
  public FooExplicit(string bar)
  {
    if (bar is null) throw new ArgumentNullException(nameof(bar));
    _bar = bar;
  }
  string IFoo.Bar => _bar;
}
```

**Reply by RikkiGibson** [reply]
The observed behavior is expected. A similar limitation exists with explicit implementations on structs. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwMQwCuKKEwKApgAQUxmUCwAUAAIBMAjM8ywMxWwEFAE4AzCAGNqASQBiAe3nMA3sypUWHAAxUAQhGFVlVAOYUEAbioBnc1YC+zR0x79rCYYQkIqC+QFEADwAHFCgJKB8QKjlFFTUNbRi/ADp9Q2MzSxs7KmcEvl9FINDwyIAKTR1gAwBKBNUmdXUoUSpymsMoayoiElqqBAALYXkAd16KCYBBYRNCMFoEADliFCCpYIQoeRhyukX5UQ662osE9XLy2Pla4e7atIMqAF4qTvOmvKdmIA==)

```cs
#nullable enable
using System;

public interface IFoo
{
  string Bar { get; set; }
}

public struct FooExplicit : IFoo
{
  string IFoo.Bar { get; set; }

  public FooExplicit(string bar) // error CS0843: Auto-implemented property 'FooExplicit.IFoo.Bar' must be fully assigned before control is returned to the caller.
  {
    if (bar is null) throw new ArgumentNullException(nameof(bar));
    ((IFoo)this).Bar = bar; // error CS0188: The 'this' object cannot be used before all of its fields have been assigned
  }
}
```

Another way you could consider working around the limitation would be to use a field initializer on the property declaration. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwMQwCuKKEwKApgAQUxmUCwAUAAIBMAjM8ywMxWwEFAE4AzCAGNqASQBiAe3nMA3sypUWHAAxUAQhGFVlVAOYUEAbioBnc1YC+zR0x792VBfICiADwAOKFASUAhUIFRyiipqGtoRngB0+obGZpY2dlT2VAC8VEQkAIQW3EzqfB6KvgFBIQAUmjrABgCUMapl6gKiVHXNhlDW+cQoLVQIABbC8gDu+RRzAILCJoRgtAgAciO+Un4IUPIwdXTr8qJ9rS0lnep1dZHyLZODLUkGuVT9N+rOzkA===)

```cs
#nullable enable
using System;

public interface IFoo
{
  string Bar { get; set; }
}

public class FooExplicit : IFoo
{
  string IFoo.Bar { get; set; } = null!;

  public FooExplicit(string bar)
  {
    if (bar is null) throw new ArgumentNullException(nameof(bar));
    ((IFoo)this).Bar = bar;
  }
}
```

**Reply by ilbrando** [reply]
I don't think this is a good recommendation. I want all my code to be typesafe and setting the property to null (which it shouldn't allow) and then silencing the compiler is not good coding style, IMHO.

**Reply by CyrusNajmabadi** [reply]
There is no way for the compiler to know this is safe.  A subclass may reimplements this interface and the assignment would go into that. The correct way is to either suppress, or control the backing field yourself (like you've done with a private field).

**Reply by CyrusNajmabadi** [reply]
> I want all my code to be typesafe

Your code is typesafe.  It is not, however perfectly null safe. But perfect null safety is a non goal for the nrt feature.  The goal is a balanced approach that provides enough values without too many false positives. In pursuit of that goal, some cracks are necessary.

**Reply by ilbrando** [reply]
> There is no way for the compiler to know this is safe.  A subclass may reimplements this interface and the assignment would go into that. The correct way is to either suppress, or control the backing field yourself (like you've done with a private field). 

I don't understand what you mean with this. If a subclass implements the interface it takes responsibility for the property and must ensure it has a value as it can't be null or am I missing something?

**Reply by ilbrando** [reply]
> > I want all my code to be typesafe
> 
> Your code is typesafe.  It is not, however perfectly null safe. But perfect null safety is a non goal for the nrt feature.  The goal is a balanced approach that provides enough values without too many false positives. In pursuit of that goal, some cracks are necessary. 

😊 I consider safe null handling part of type safety and I want all my code to be safe. If I'm forced to use things like `= null!` I feel I'm doing something wrong or the compiler has some shortcomings.

**Reply by CyrusNajmabadi** [reply]
> I don't understand what you mean with this. If a subclass implements the interface it takes responsibility for the property and must ensure it has a value as it can't be null or am I missing something?

I'm saying that the assignment in *your* constructor will call into the subclass, and the backing field for *your* explicit interface will remain null and uninitialized.  That's why the solution here is to use an actual property initializer as that's the only thing guaranteed to write into *that* property's backing field.

> 😊 I consider safe null handling part of type safety

It is not, and we do not. 

> and I want all my code to be safe. 

You can write your own analyzers/suppressors to have this logic then.  That is a stated *non goal* of NRT.  

> I feel I'm doing something wrong

You're not doing something wrong.  You just have expectations that are counter to our explicit design goals for this feature.

> or the compiler has some shortcomings.

Yes.  The compiler has many limitations here.  That's necessary as the analysis for this sort of problem is unbounded in scope, and extremely expensive.  As such, we've made certain decisions to make sure we can do this work in a cost effective fashion while balancing the need for good true positives, and low incidences of false positives.

If you are ok with any amount of expense to get your definition of 'type safety', then that avenue is available with analyzers and suppressors.

**Reply by ilbrando** [reply]
I'm sorry, but i don't know what NRT means and I'm only trying to give my input on what I find to be good code quality and how I hope the compiler helps me in writing it. I have used TypeScript a lot for several years now and continue to be amazed at how they continue to improve type safety (in the way I understand the word).

**Reply by CyrusNajmabadi** [reply]
> i don't know what NRT means

"Nullable reference types".  The feature here that is saying "you declared this as non-null, but it may still be null at runtime".

>  and I'm only trying to give my input on what I find to be good code qualit

RIght.  But we're telling you: write it in the following way. :)

You can choose to reject our suggestions, but then we're at an impasse.

>  I have used TypeScript a lot for several years now and continue to be amazed at how they continue to improve type safety (in the way I understand the word).

TS is much less typesafe than C#.  Allowing, for example, types to be entirely different from what they are stated at compile time whne actually executing at runtime.  

Our NRT impl is very similar to TS's similar approach to giving or avoiding null warnings.  However, TS also doesn't have this concept of explicit interface implementation, so the comparison doesn't work here :)

**Reply by ilbrando** [reply]
I think this is going a bit in a wrong direction. I still think my original question is valid. Everybody can see the property is initialized in the constructor - except the compiler. It can be solved with a private field, but I don't need/want a private field, so this is kind of a work around, and it would be better if the compiler could infer what is happening exactly as it can in the implicit version.

**Reply by CyrusNajmabadi** [reply]
> Everybody can see the property is initialized in the constructor - except the compiler. 

It is not. The constructor is casting 'this' to an interface, which can and will write into some other property. 

> but I don't need/want a private field, 

Then you can suppress the warning. We provide the tooling for a user to say: I know better about the entirety of my code, and there's no way this will ever write into something else, so please be quiet. :-)
